microsoft/testsuites/cloud_hypervisor/ch_tests_tool.py

```diff
@@ -0,0 +1,98 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+import re
+from pathlib import PurePath
+from typing import Any, List, Optional, Type
+
+from lisa.executable import Tool
+from lisa.operating_system import CBLMariner
+from lisa.tools import Docker, Echo, Git, Modprobe
+
+
+class CloudHypervisorTests(Tool):
+    TIME_OUT = 1800
+
+    repo = "https://github.com/cloud-hypervisor/cloud-hypervisor.git"
+
+    cmd_path: PurePath
+    repo_root: PurePath
+
+    @property
+    def command(self) -> str:
+        return str(self.cmd_path)
+
+    @property
+    def can_install(self) -> bool:
+        return True
+
+    @property
+    def dependencies(self) -> List[Type[Tool]]:
+        return [Git, Docker]
+
+    def run_tests(self, test_type: str, skip: Optional[List[str]] = None) -> List[str]:
+        self.node.tools[Modprobe].load("openvswitch")
+
+        if skip is not None:
+            skip_args = " ".join(map(lambda t: f"--skip {t}", skip))
+        else:
+            skip_args = ""
+
+        result = self.run(
+            f"tests --{test_type} -- -- {skip_args}",
+            timeout=self.TIME_OUT,
+            force_run=True,
+            cwd=self.repo_root,
+            no_info_log=False,  # print out result of each test
+        )
+
+        failures = self._extract_failed_tests(result.stdout)
+        if not failures:
+            result.assert_exit_code()
+
+        return failures
+
+    def _initialize(self, *args: Any, **kwargs: Any) -> None:
+        tool_path = self.get_tool_path(use_global=True)
+        self.repo_root = tool_path / "cloud-hypervisor"
+        self.cmd_path = self.repo_root / "scripts" / "dev_cli.sh"
+
+    def _install(self) -> bool:
+        git = self.node.tools[Git]
+        git.clone(self.repo, self.get_tool_path(use_global=True))
+        if isinstance(self.node.os, CBLMariner):
+            daemon_json_file = PurePath("/etc/docker/daemon.json")
+            daemon_json = '{"default-ulimits":{"nofile":{"Hard":65535,"Name":"nofile","Soft":65535}}}'  # noqa: E501
+            self.node.tools[Echo].write_to_file(
+                daemon_json, daemon_json_file, sudo=True
+            )
+        self.node.tools[Docker].start()
+
+        return self._check_exists()
+
+    def _extract_failed_tests(self, output: str) -> List[str]:
+        # The failures list is output by the test runner in this format:
+        #
+        # failures:
+        #     failed_test_name_1
+        #     failed_test_name_2
+        #     ... so on
+        #
+        # To parse, we first find the "failures:" line and then parse the
+        # following lines that begin with one or more whitespaces to extract
+        # the test name.
+        lines = output.split("\n")
```

Comment:
I only see below line in output.
```test result: ok. 10 passed; 0 failed; 10 ignored; 0 measured; 105 filtered out; finished in 994.68s```

Reply:
We don't need to parse this line. When there are failures, it will print the list of failed tests. That's what we are trying to parse here.